wai-extra/Network/Wai/Test.hs

```diff
@@ -214,8 +216,9 @@ assertString :: HasCallStack => String -> Session ()
 assertString s = unless (null s) $ assertFailure s
 
 assertFailure :: HasCallStack => String -> Session ()
-assertFailure msg = msg `deepseq` liftIO (throwIO (WaiTestFailure msg))
+assertFailure = liftIO . HUnit.assertFailure
```

Comment:
This may just be deferring the breakage until runtime. If anyone is checking for `WaiTestFailure` in their code, this will now silently break. Is there no way to keep the call stack _and_ keep `WaiTestFailure`?

Reply:
Yes, that is true.  A user will still get a deprecation warning, so it's not 100% silent, but still not ideal.

We can try to turn `WaiTestFailure` into a pattern synonym; I'll have to check for compiler support, don't know from the top of my head when they were introduced.

Friday night here already, I'll look into this tomorrow.

Reply:
> I'll have to check for compiler support

Pattern synonyms were introduced with GHC `7.8.1`.  We require `base >= 4.8`, which implies GHC `7.10.1` or later.  So this should work.

Reply:
> Is there no way to keep the call stack and keep `WaiTestFailure`?

I'm not aware of a backwards compatible and straight forward way, no.

If we would want to keep `WaiTestFailure` as a data type and add source locations to it, we would need to change it to something like:
```haskell
data WaiTestFailure = WaiTestFailure SrcLoc String
```